refactor: optimize ConvertTo-QueryString with better null handling

SoufiyanAk opened the pull request:
- Add null input validation to prevent errors
- Improve null value handling for hashtable entries and object properties
- Optimize performance by reducing redundant Get-Member calls
- Enhance error messages with more context about supported types
- Add proper ToString() conversion for non-null values

merill commented:
@SoufiyanAk Since this is a core part of Maester I want to be extra careful that we don't cause any issues.

I've added some Pester tests in #1042

Would you be able to add additional tests for the changes you are making? 

Can I know what prompted this PR? 

Where there specific calls to Invoke-GraphRequest that were failing? If so, can you include them here in the comments and ideally as a pester test so we can verify the changes?

Also this test is doing a few different things in addition to the null check (eg. perf etc). Might be better to split them into different ones.

We are usually not this pedantic, but since this cmdlet is a core part of Maester and affects every test we need to be extra careful. Tx and much appreciated.

merill commented:
@SoufiyanAk I'm going to move this to draft. Would you be able to reply to some of the queries raised? Tx.

SoufiyanAk commented:
Thanks for the thorough review. I totally understand being extra careful with core Maester components.
I've added the additional tests you asked for, split into separate categories as you suggested:
Main test file (ConvertTo-QueryString.Tests.ps1):

- Null handling tests (null inputs, hashtable null values, object property nulls)
- Performance tests (execution time validation, Get-Member optimization)
- Error handling tests (enhanced error messages, proper categorization)
- Integration tests (real-world Graph API scenarios)

Specific failure scenarios (ConvertTo-QueryString.Invoke-GraphRequest.Tests.ps1):

- The exact scenarios that were causing issues with Graph requests
- Real-world examples like user queries, group membership queries, etc.

What prompted this PR:
I was running into issues where Invoke-GraphRequest would fail when processing query parameters with null values. Stuff like:

> $queryParams = @{
>     '$select' = 'id,displayName,mail'
>     '$filter' = $null  # This was causing problems
>     '$top' = 100
>     '$skip' = $null    # This too
> }

The original function would choke on these null values and either throw errors or produce malformed query strings. I also noticed it was making redundant Get-Member calls which was slowing things down, especially with objects that had lots of properties so i added proper null handling so null values become empty strings instead of causing errors
and i optimized the Get-Member calls for better performance, i improved error messages to be more helpful when debugging and i added the EncodeParameterNames parameter for URL encoding

I've tested it with PowerShell and the core functionality works correctly. The tests validate that null handling works as expected and performance is improved.

Let me know if you need anything else or have other concerns!

merill commented:
@SoufiyanAk Thanks a lot for the updates!

A quick note. I just realized the test files you added are not in the right folder.

They need to be in ./powershell/tests/functions along with the rest of the tests for internal cmdlets. BTW I already added one file along with tests at ./powershell/tests/functions/ConvertTo-QueryString.Tests.ps1

Can you please add your tests into the same file and also validate they are passing by running ./powershell/tests/pester.ps1

I tried copying your tests but over 25 of them are failing atm.

Let me know if you have any queries.
